third_party/nvfuser/csrc/scheduler/normalization.cpp

```diff
@@ -1,35 +1,232 @@
+      dev_prop->warpSize * dev_prop->warpSize;
+  int64_t blocks_per_sm =
+      std::min(blocks_per_sm_max1, threads_per_sm / threads_per_block);
+  int64_t gdimy = blocks_per_sm * device_multiprocessor_count;
+  int64_t bdimx = scheduler_utils::roundUpPow2(ceilDiv(after_unroll, gdimy));
+  int64_t bdimy = threads_per_block / bdimx;
+
+  // if gdimy is too large, reduce it and increase bdimx, decrease bdimy, this
+  // may increase wave tailings
+  const int64_t outer_iter_min = 8;
+  const int64_t gdimy_max = scheduler_utils::roundUpToN(
+      outer_dim_numel / outer_iter_min, device_multiprocessor_count);
+  while (gdimy > gdimy_max && blocks_per_sm > 1) {
+    blocks_per_sm -= 1;
+    gdimy = blocks_per_sm * device_multiprocessor_count;
+    bdimx = scheduler_utils::roundUpToN(ceilDiv(after_unroll, gdimy), 2);
+    while (threads_per_block % bdimx) {
+      bdimx += 2;
+    }
+    bdimy = threads_per_block / bdimx;
+  }
 
-// round up to multiple of 8 or pow2 whichever smaller
-int64_t roundUpPow2Or8(const int64_t x) {
-  auto round_up_pow2 = scheduler_utils::lastPow2(x);
-  if (round_up_pow2 < x) {
-    round_up_pow2 *= 2;
+  // tuned for layer_norm backward on A100
+  // when inner_dim_numel <= 1024, bdimx is usually small after divide by
+  // vectorization factor and batch of persistent buffers then bdimy is used to
+  // parallelize outer_dim instead of inner_dim
+  if (inner_dim_numel <= 1024) {
+    rparams->multiple_reds_per_blk = true;
+    rparams->tidx_for_outer_reduction = true;
+    threads_per_block_max = 512;
+
+    // step-1, inner_dim_numel is reduction dim of inner reduction
+    batch = 1;
+    vect = inner_dim_numel > dev_prop->warpSize ? vectorize_factor : 1;
+    bdimx = inner_dim_numel / vect;
+
+    // step-2, inner_dim_numel is iteration dim of outer reduction
+    unroll_factor = std::min(vect, (int64_t)2);
+    after_unroll = ceilDiv(inner_dim_numel, unroll_factor);
+    blocks_per_sm = outer_dim_numel > device_multiprocessor_count ? 2 : 1;
+    gdimy = blocks_per_sm * device_multiprocessor_count;
+    bdimy = std::min(
+        ceilDiv(after_unroll, gdimy), ceilDiv(threads_per_block_max, bdimx));
+    rparams->block_dim_iter_dom = ParallelType::TIDy;
+    // TIDx is used for both inner and outer reduction
+    if (bdimx % dev_prop->warpSize == 0) {
+      rparams->pad_inner_reduction_to_warp = true;
+      rparams->pad_outer_reduction_to_warp = true;
+    }
+
+  } else {
+    rparams->block_dim_inner_reduction_extra = ParallelType::TIDy;
   }
-  constexpr int64_t kEight = 8; // clang tidy
-  auto round_up_8 = x % kEight == 0 ? x : x + (kEight - x % kEight);
-  return std::min(round_up_8, round_up_pow2);
+
+  reg_per_thread =
+      getRegPerThreadGivenThreadsPerSM(bdimx * bdimy * blocks_per_sm);
+  rparams->vectorization_factor_tmp_gmem_read =
+      std::min((int64_t)4, unroll_factor);
+  ;
+  rparams->vectorization_factor_tmp_gmem_write = std::min((int64_t)4, vect);
```

Comment:
When `vect` is 8, this means the writes to the tmp buffer is vectorized by 4. Does it actually work? I wonder if it needs to be vectorized by the same degree.

Reply:
Same here